File: web/src/main/java/hu/bsstudio/robonaut/safetycar/SafetyCarOvertakeHandler.java
package hu.bsstudio.robonaut.safetycar;

import hu.bsstudio.robonaut.safetycar.model.SafetyCarOvertakeInformation;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class SafetyCarOvertakeHandler implements HandlerFunction<ServerResponse> {

    @NonNull
    private final SafetyCarService service;

    @Override
    public Mono<ServerResponse> handle(final ServerRequest request) {
        final var detailedTeam = request.bodyToMono(SafetyCarOvertakeInformation.class)
            .flatMap(service::safetyCarWasOvertaken);
        return ServerResponse.ok().body(detailedTeam, DetailedTeam.class);
    }
}

File: web/src/main/kotlin/hu/bsstudio/robonaut/safetycar/SafetyCarOvertakeHandler.kt
package hu.bsstudio.robonaut.safetycar

import hu.bsstudio.robonaut.safetycar.model.SafetyCarOvertakeInformation
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono

class SafetyCarOvertakeHandler(private val service: SafetyCarService) : HandlerFunction<ServerResponse> {

    override fun handle(request: ServerRequest): Mono<ServerResponse> {
        return request.bodyToMono<SafetyCarOvertakeInformation>()
            .flatMap(service::safetyCarWasOvertaken)
            .let { detailedTeam -> ServerResponse.ok().body(detailedTeam) }
    }
}
